govuk_chat_evaluation/rag_answers/data_models.py

```diff
@@ -10,11 +33,108 @@ class GenerateInput(BaseModel):
+    @classmethod
+    def flatten_retrieved_context(cls, data: dict[str, Any]) -> dict[str, Any]:
+        """Convert the retrieved context from a list of dictionaries to a flatnlist of strings. 
+        This is done to ensure compatibility with the LLMTestCase class and for better compatibility with the Faitfulness metric."""
+        raw_context = data.get("retrieved_context", [])
+        if raw_context and isinstance(raw_context[0], dict):
+
+            structured = []
+            for i, ctx_dict in enumerate(raw_context):
+                try:
+                    structured_ctx = StructuredContext(**ctx_dict)
+                    structured.append(structured_ctx)
+                except Exception as e:
+                    raise ValueError(f"Invalid context at index {i}: {e}")
+                
+            data["retrieved_context"] = [
+                f"{ctx.title}\n"
+                f"{' > '.join(ctx.heading_hierarchy)}\n"
+                f"{ctx.description}\n\n"
+                f"{ctx.html_content}"
+                for ctx in structured
+            ]
+        return data
+
+
+class MetricName(str, Enum):
+    FAITHFULNESS = "faithfulness"
+    RELEVANCE = "relevance"
+    BIAS = "bias"
+    # others to add
+
+class LLMJudgeModel(str, Enum):
+    GPT_4O_MINI = "gpt-4o-mini"
+    GPT_4O = "gpt-4o"
+    AMAZON_NOVA_MICRO_1 = "eu.amazon.nova-micro-v1:0"
+    AMAZON_NOVA_PRO_1 = "eu.amazon.nova-pro-v1:0"
+    GEMINI_15_PRO = "gemini-1.5-pro-002"
+    GEMINI_15_FLASH = "gemini-1.5-flash-002"
+
+
+class LLMJudgeModelConfig(BaseModel):
+    model: LLMJudgeModel
+    temperature: float = 0.0
+    
+    def instantiate_llm_judge(self):
+        """Return the LLM judge model instance."""
+        match self.model:
+            case LLMJudgeModel.AMAZON_NOVA_MICRO_1:
+                return None  # Placeholder for actual class instance
+            case LLMJudgeModel.AMAZON_NOVA_PRO_1:
+                return None  # Placeholder
+            case LLMJudgeModel.GEMINI_15_PRO:
+                return None  # Placeholder
+            case LLMJudgeModel.GEMINI_15_FLASH:
+                return None  # Placeholder
+            case LLMJudgeModel.GPT_4O_MINI | LLMJudgeModel.GPT_4O:
+                return self.model.value  # Just returns the string name as they are in-built models for DeepEval llm judge
+    
+
+class MetricConfig(BaseModel):
+    name: MetricName
+    threshold: float
+    # model: str | DeepEvalBaseLLM
+
+    def to_metric_instance(self, llm_judge: DeepEvalBaseLLM | None = None):
+        match self.name:
+            case MetricName.FAITHFULNESS:
+                return FaithfulnessMetric(threshold=self.threshold, model=llm_judge)
+            case MetricName.RELEVANCE:
+                return AnswerRelevancyMetric(threshold=self.threshold, model=llm_judge)
+            case MetricName.BIAS:
+                return BiasMetric(threshold=self.threshold, model=llm_judge)
+
+
+class EvaluationConfig(BaseModel):
```

Comment:
I am a bit confused about this and the related models. As per the comment below it sounds like you'd like to improve them too so perhaps we could have a chat about the individual needs and roles to iterate their modelling.